feat: save COGs with dask to azure

wietzesuijker opened the pull request:
This PR introduces Azure Blob Storage support for saving COGs with Dask, complementing the existing AWS S3 support.

- Implements a MultiPartUpload class in _az.py to handle Azure-specific multipart upload operations, including block staging, finalization, and cancellation.
- Adds MultiPartUploadBase in _multipart.py to provide a shared interface for multipart uploads, applicable to both Azure and S3 implementations.
- Updates save_cog_with_dask to recognize Azure URLs and integrate the Azure multipart upload workflow.
- Adds unit tests in tests/test_az.py using mocked Azure SDK to verify functionality specific to Azure Blob Storage.

Here’s a simple test I ran to verify the Azure implementation:

```python
import dask.distributed
import planetary_computer as pc
from odc.geo.cog._tifffile import save_cog_with_dask
from odc.stac import configure_rio, stac_load
from pystac_client import Client

# Setup Dask client for efficient processing
client = dask.distributed.Client()
configure_rio(cloud_defaults=True, client=client)

# Query STAC API for Sentinel-2 data
catalog = Client.open("https://planetarycomputer.microsoft.com/api/stac/v1")
query = catalog.search(
    collections=["sentinel-2-l2a"],
    datetime="2019-06",
    query={"s2:mgrs_tile": dict(eq="06VVN")},
)
items = list(query.items())
print(f"Found {len(items)} datasets")

# Load Sentinel-2 data with specific bands
resolution = 40 
xx = stac_load(
    items,
    bands=["SCL"],
    resolution=resolution,
    chunks={"x": 2048, "y": 2048},
    patch_url=pc.sign,
    dtype="uint16",
    nodata=0,
)

# Select a single time slice for saving
to_save = xx.SCL.isel(time=3)

# Azure Blob Storage details
blob_name = f"SCL-{to_save.time.dt.strftime('%Y%m%d').item()}.tif"
azure_url = f"az://{container}/{blob_name}"


result = save_cog_with_dask(
    to_save,
    dst=f"az://{container}/{blob_name}",
    compression="DEFLATE",
    predictor=2,
    overview_resampling="nearest",
    azure={
        "account_url": f"https://{storage_account}.blob.core.windows.net",
        "credential": sas_token,  # can be blob specific or container wide
    },
    blocksize=[512, 512],
)

# Compute the result to trigger the actual upload
result.compute()
print(f"COG successfully uploaded to: {azure_url}")
```

While implementing this, I noticed a couple of quirks in the original code:

- Spelling Inconsistencies: Words like finalise vs. finalize were used. I  used **`ise`** rather than `ize`.
- Typing Updates: The code inconsistently uses Dict/List from typing vs. built-ins (dict/list). I've changed it to Dict and List since the dev-env.yml specifies python=3.8 (but I hadn't used python 3.8 in ages).

As this is my first PR here, I look forward to feedback and suggestions for improvement!

Kirill888 commented:
@wietzesuijker thanks for the PR.

Can I ask you to undo `dict -> Dict`, `list -> List` changes, we no longer aim to support 3.8 going forward as most of our dependencies do not support it either. This also blows up review effort. When making changes like this please try to separate "clean up changes" from "actual changes" and use separate commits for each separate logical unit of work, example:

- fix spelling
- mypy harmonization
- add doc comments
- implement azure back-end


I'll need to find some time to review this properly and to remember what was the original plan for extending to other blob stores. At this point I'm not convinced we need a base class, but my memory of this code is hazy. In the meantime please

- undo mypy changes
- separate changes to doc comments from the "real change"
- make sure to run `black` formatter, or do the pre-commit thing
- push new change history into this PR with `--force-with-lease` option

Kirill888 commented on odc/geo/cog/_s3.py:
```diff
@@ -126,31 +129,33 @@ def write_part(self, part: int, data: SomeData) -> dict[str, Any]:
-    def finalise(self, parts: list[dict[str, Any]]) -> str:
+    def finalise(self, parts: List[Dict[str, Any]]) -> str:
+        """Finalise the multipart upload."""
+        # pylint: disable=import-outside-toplevel,import-error
```
pylint flags?

wietzesuijker replied:
oops

Kirill888 commented on tests/test_az.py:
```diff
@@ -0,0 +1,62 @@
+from odc.geo.cog._az import MultiPartUpload
+
+
+def test_s3_mpu():
```
test_s3_mpu cut and paste name?

wietzesuijker replied:
oops 2

wietzesuijker committed:
Feat: MultiPartUploadBase class for multipart upload

wietzesuijker committed:
Feat: azure backend to save cogs with dask

wietzesuijker commented:
Thanks @Kirill888; I addressed your suggestions.

If you're not too familiar with Azure and want to give it a spin, you can:
- Sign up at [Azure Free](https://azure.microsoft.com/en-us/free/).
- In [Azure Portal](https://portal.azure.com/), create a storage account and container.
- Generate a SAS token in the storage account's Shared access signature settings.
- Use the storage account URL and SAS token in the PR tests.
- Delete the resource group after testing.

Thanks again, and great to have your input on this!

Kirill888 commented on odc/geo/cog/_tifffile.py:
```diff
@@ -10,19 +10,22 @@
 from ..geobox import GeoBox
 from ..math import resolve_nodata
 from ..types import Shape2d, SomeNodata, Unset, shape_
+from ._az import MultiPartUpload as AzMultiPartUpload
```
this will raise an error if `azure` is not installed, we should be able to save to S3 even then though.

Also might as well name that class same way in `_az.py`.

wietzesuijker replied:
Fair! I've addressed that. I didn't dare to rename the S3MultiPartUpload in _s3.py before. I've also added HAVE_S3 and HAVE_AZURE flags to take care of non-installed packages if either is used. Not sure if you're a fan of that.

Kirill888 commented on odc/geo/cog/_az.py:
```diff
@@ -0,0 +1,191 @@
+import base64
+from typing import Any, Union
+
+from azure.storage.blob import BlobBlock, BlobServiceClient
```
move this import into methods that need that (and tell pylint it's ok). We need to be able to import from here even when `azure` is not installed in the environment.

wietzesuijker replied:
I tried, but I couldn't get `python -m unittest tests/test_az.py` to behave without the import being outside of the class. Would you know how?

Kirill888 commented on tests/test_az.py:
```diff
@@ -0,0 +1,74 @@
+from unittest.mock import MagicMock, patch
+
+from odc.geo.cog._az import MultiPartUpload
+
```
skip this test if `azure` is not available.

wietzesuijker replied:
Added the HAVE_AZURE flag here to be used as decorator @require_azure

wietzesuijker committed:
Feat: safely import azure-storage-blob and boto3

Kirill888 commented:
### Optional Dependency Handling

Example:

https://github.com/opendatacube/odc-geo/blob/62b7c5ebe41a3c7fc19f267daf06246e437ff830/odc/geo/_xr_interop.py#L55-L61 

The `have` object is here:

https://github.com/opendatacube/odc-geo/blob/62b7c5ebe41a3c7fc19f267daf06246e437ff830/odc/geo/_interop.py#L22-L28

Add `azure` to that, then use it instead of try imports.

### Tests

We use `pytest` for tests, here is example of skipping whole file when needed dependencies are not present

https://github.com/opendatacube/odc-geo/blob/62b7c5ebe41a3c7fc19f267daf06246e437ff830/tests/test_converters.py#L6-L9

### Packaging

Add optional `azure` feature here:

https://github.com/opendatacube/odc-geo/blob/62b7c5ebe41a3c7fc19f267daf06246e437ff830/setup.cfg#L54-L60


### Run checks locally

`mypy` and `pylint` checks are failing, run them locally or setup actions in your fork, actions on the PR require manual approval  from maintainers.

### Simplifying Review process

There is now quite a bit of cleanup changes here, let's have those in a separate PR that I can merge, then we can focus on actual change without getting bogged down by all the mypy harmonization changes.

wietzesuijker commented:
Added [a PR](https://github.com/opendatacube/odc-geo/pull/196) for the cleanup. I'll give the rest another push tomorrow (UK time). Hopefully, that will give us enough time to review before the holiday season; if not, we can review again in 2025 :).

Kirill888 commented:
@wietzesuijker other PR is merged, and some other changes, please rebase this against main.

wietzesuijker committed:
Feat: allow s3 and or az dependencies

Merged branch 'develop' into feat/save-cog-to-azure.

wietzesuijker commented:
Thanks @Kirill888! I didn't get as much done as I'd liked. I updated the test for az and started using `have.azure` and `have.s3`. Could you review the code to see if I'm on the right track? I might need to git rebase and rewrite the history, but I pushed that issue into tomorrow as I was working on this when it wasn't merged yet.
Thanks again.

Kirill888 commented on odc/geo/_interop.py:
```diff
@@ -43,6 +43,14 @@ def datacube(self) -> bool:
     def tifffile(self) -> bool:
         return self._check("tifffile")
 
+    @property
+    def azure(self) -> bool:
+        return self._check("azure.storage.blob")
+
+    @property
+    def s3(self) -> bool:
+        return self._check("boto3")
+
```
- we use lower level `botocore` library
- property name is library name, not feature
- `s3 -> botocore` and `boto3 -> botocore`

wietzesuijker replied:
fair, changed that to botocore, noting that boto3 is in the requirements.

Kirill888 commented on tests/test_az.py:
```diff
@@ -1,96 +1,80 @@
-except ImportError:
-    AzMultiPartUpload = None
-    HAVE_AZURE = False
+pytest.importorskip("odc.geo.cog._az")
+from odc.geo.cog._az import AzMultiPartUpload  # noqa: E402
```
This won't work as expected as `odc.geo.cog._az`CAN be imported now even when `azure` is missing, it just won't function after the import. I guess it should be `pytest.importorskip("azure.storage.blob")`

wietzesuijker replied:
thanks :), changed

Kirill888 commented:
@wietzesuijker there is currently a conflict after merging your other PR, can you please rebase this branch against `develop` branch: `git pull --rebase upstream/develop` or something along these lines.

Kirill888 committed:
Fix: mypy errors in geo/cog modules

wietzesuijker committed:
Feat: Refactor upload method and reduce duplication

- Restored type hints in `upload` for improved type safety.
- Added `get_mpu_kwargs` to centralize shared keyword arguments.
- Simplified `upload` and `mpu_upload` implementations by reusing `get_mpu_kwargs`.
- Reduced code duplication across `_mpu.py` and `_multipart.py`.

wietzesuijker commented:
@Kirill888 I've added some changes to fix non-related mypy errors in https://github.com/opendatacube/odc-geo/pull/195/commits/3b0223c16bbf6459f3f12e39fcbe25b491572b8a and squashed that with your commit.
Also, pylint complained about duplicate code, so I pushed https://github.com/opendatacube/odc-geo/pull/195/commits/e1b6187771bbb1a2413624ef9a43d46c82bbf17f.

The tests pass (after a few tries), and the basic example in the description of the PR still works. 
If you have any other suggestions, let me know.

Kirill888 reviewed (approved):
Thanks and sorry for the delay (holidays, and it's summer here too).
